### PR TITLE
scm extension point - {1,2,3,4,5}/7

### PR DESCRIPTION
**Pull request by srinivasupadhya**
This PR takes care of 1, 2, 3, 4 & 5 in #818. VSM material node display (image for pluggable SCM) needs to be taken care.

### CODE REVIEW

**Comment by srinivasupadhya** [comment]
Use [this](https://github.com/srinivasupadhya/git-poller) sample plugin for testing end-poing.

File: agent/src/com/thoughtworks/go/agent/AgentController.java [hunk 1]
```diff
@@ -62,13 +64,17 @@
...
     private AgentUpgradeService agentUpgradeService;
+    private PackageAsRepositoryExtension packageAsRepositoryExtension;
+    private SCMExtension scmExtension;
     private TaskExtension taskExtension;
 
     @Autowired
     public AgentController(BuildRepositoryRemote server, GoArtifactsManipulator manipulator, SslInfrastructureService sslInfrastructureService, AgentRegistry agentRegistry,
                            AgentUpgradeService agentUpgradeService, SubprocessLogger subprocessLogger, SystemEnvironment systemEnvironment,
-                           PluginManager pluginManager, TaskExtension taskExtension) {
+                           PluginManager pluginManager, PackageAsRepositoryExtension packageAsRepositoryExtension, SCMExtension scmExtension, TaskExtension taskExtension) {
```
Comment (arvindsv) [inline agent/src/com/thoughtworks/go/agent/AgentController.java]: Should we have ```Extensions``` or something? I don't see the need to change this every time we add a new kind of extension. Could use ```extensions.forTask()``` and ```extensions.forPackageRepository()```, etc.

Comment (arvindsv) [inline agent/src/com/thoughtworks/go/agent/AgentController.java]: Anyway, all that is being done here is that it is passed down to the runner.

File: agent/src/com/thoughtworks/go/agent/JobRunner.java [hunk 2]
```diff
@@ -45,10 +47,10 @@ private void cancelJob(AgentRuntimeInfo agentRuntimeInfo) {
         handled = true;
     }
 
-    public void run(Work work, AgentIdentifier agentIdentifier, BuildRepositoryRemote server,
-                    GoArtifactsManipulator manipulator, AgentRuntimeInfo agentRuntimeInfo, TaskExtension taskExtension) {
+    public void run(Work work, AgentIdentifier agentIdentifier, BuildRepositoryRemote server, GoArtifactsManipulator manipulator, AgentRuntimeInfo agentRuntimeInfo,
+                    PackageAsRepositoryExtension packageAsRepositoryExtension, SCMExtension scmExtension, TaskExtension taskExtension) {
```
Comment (arvindsv) [inline agent/src/com/thoughtworks/go/agent/JobRunner.java]: Same. Extensions class would help here as well.

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+import static org.apache.commons.lang.StringUtils.upperCase;
+
+public class PluggableSCMMaterial extends AbstractMaterial {
+    public static final String TYPE = "PluggableSCMMaterial";
+
+    private String scmId;
+
+    @Expose
+    @SerializedName("scm")
+    private SCM scmConfig;
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: Can you think of a more descriptive name than SCM? Maybe PluggedSCMConfig or even SCMConfig?

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+    public void setFolder(String folder) {
+        this.folder = folder;
+    }
+
+    public Filter getFilter() {
+        return filter;
+    }
+
+    public void setFilter(Filter filter) {
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: Why are these methods needed? Ibatis/Hibernate?

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        this.filter = filter;
+    }
+
+    public String getPluginId() {
+        return getScmConfig().getPluginConfiguration().getId();
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: Can getScmConfig() not be null here? I ask because you're checking for it just below in getFingerprint().

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+import static com.thoughtworks.go.util.command.EnvironmentVariableContext.escapeEnvironmentVariable;
+import static java.lang.String.format;
+import static org.apache.commons.lang.StringUtils.isEmpty;
+import static org.apache.commons.lang.StringUtils.upperCase;
+
+public class PluggableSCMMaterial extends AbstractMaterial {
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: What is the reason for it being an AbstractMaterial and not an SCMMaterial?

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        // skipping super.equals()
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: Why?

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = super.hashCode();
+        result = 31 * result + (scmId != null ? scmId.hashCode() : 0);
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: Very dangerous territory. It's almost always a mistake to keep equals() and hashCode() different like this. I know it's done the same way in PackageMaterial, but this should be reasoned about, and if confident that it cannot cause issues when this object is in a hash table, then we should comment on top of this method and the corresponding one in PackageMaterial.

File: common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        return scmPropertyConfiguration;
+    }
+
+    private void populateConfiguration(Configuration configuration, com.thoughtworks.go.plugin.api.config.Configuration pluginConfiguration) {
+        for (ConfigurationProperty configurationProperty : configuration) {
+            pluginConfiguration.add(new SCMProperty(configurationProperty.getConfigurationKey().getName(), configurationProperty.getValue()));
+        }
+    }
+
+    private File workingDirectory(File baseFolder, String destinationFolder) {
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java]: Why not do it like with [ScmMaterial](https://github.com/gocd/gocd/blob/master/common/src/com/thoughtworks/go/domain/materials/MaterialAgentFactory.java#L49)? Meaning, this can be in the Material itself, rather than here.

File: common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Override
+    public void prepare() {
+        PluggableSCMMaterial material = (PluggableSCMMaterial) revision.getMaterial();
+        Modification latestModification = revision.getLatestModification();
+        SCMRevision scmRevision = new SCMRevision(latestModification.getRevision(), latestModification.getModifiedTime(), null, null, latestModification.getAdditionalDataMap(), null);
+        File destinationFolder = workingDirectory(workingDirectory, material.getFolder());
+        Result result = scmExtension.checkout(material.getScmConfig().getPluginConfiguration().getId(), buildSCMPropertyConfigurations(material.getScmConfig()), destinationFolder.getAbsolutePath(), scmRevision);
+        if (!result.isSuccessful()) {
+            // handle
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java]: Hmm. Needs to be done, right?

File: common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        }
+        // handle messages
+    }
+
+    private SCMPropertyConfiguration buildSCMPropertyConfigurations(SCM scmConfig) {
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java]: Shouldn't this be the responsibility of the ScmExtension class? Should probably ask @pshivana.

File: common/src/com/thoughtworks/go/remote/work/BuildWork.java [hunk 12]
```diff
@@ -78,14 +80,13 @@ private void initialize(BuildRepositoryRemote remoteBuildRepository,
...
-                       GoArtifactsManipulator goArtifactsManipulator, EnvironmentVariableContext environmentVariableContext,
-                       AgentRuntimeInfo agentRuntimeInfo, TaskExtension taskExtension) {
+    public void doWork(AgentIdentifier agentIdentifier, BuildRepositoryRemote remoteBuildRepository, GoArtifactsManipulator goArtifactsManipulator,
+                       EnvironmentVariableContext environmentVariableContext, AgentRuntimeInfo agentRuntimeInfo,
+                       PackageAsRepositoryExtension packageAsRepositoryExtension, SCMExtension scmExtension, TaskExtension taskExtension) {
         initialize(remoteBuildRepository, goArtifactsManipulator, agentRuntimeInfo, taskExtension);
         environmentVariableContext.addAll(assignment.initialEnvironmentVariableContext());
         try {
-            JobResult result = build(environmentVariableContext, agentIdentifier);
+            JobResult result = build(environmentVariableContext, agentIdentifier, packageAsRepositoryExtension, scmExtension);
```
Comment (arvindsv) [inline common/src/com/thoughtworks/go/remote/work/BuildWork.java]: An Extensions class, again, will help.

File: config/config-api/src/com/thoughtworks/go/config/materials/MaterialConfigs.java [hunk 13]
```diff
@@ -289,12 +318,15 @@ public void setConfigAttributes(Object attributes) {
...
+        } else if (PluggableSCMMaterialConfig.TYPE.equals(materialType)) {
+            addMaterialConfig(getSCMMaterial(), (Map) attributeMap.get(PluggableSCMMaterialConfig.TYPE));
         }
     }
 
     public boolean scmMaterialsHaveDestination() {
-        for (ScmMaterialConfig scmMaterial : filterScmMaterials()) {
-            if (!scmMaterial.hasDestination()) {
+        for (MaterialConfig scmMaterial : getSCMAndPluggableSCMConfigs()) {
+            String destination = (scmMaterial instanceof ScmMaterialConfig) ? ((ScmMaterialConfig) scmMaterial).getFolder() : ((PluggableSCMMaterialConfig) scmMaterial).getFolder();
```
Comment (arvindsv) [inline config/config-api/src/com/thoughtworks/go/config/materials/MaterialConfigs.java]: The MaterialConfig interface has a getFolder() method. Why is this needed? The condition check, I mean.

File: config/config-api/src/com/thoughtworks/go/config/materials/MaterialConfigs.java [hunk 14]
```diff
@@ -155,22 +149,43 @@ public void validate(ValidationContext validationContext) {
...
+        allSCMMaterials.addAll(scmMaterials);
+        allSCMMaterials.addAll(pluggableSCMMaterials);
+        return allSCMMaterials;
+    }
+
+    private void validateDestinationFolder(List<MaterialConfig> allSCMMaterials, MaterialConfig material) {
+        String materialFolder = material.getFolder();
+        for (MaterialConfig otherMaterial : allSCMMaterials) {
+            if (otherMaterial != material) {
+                if (otherMaterial instanceof ScmMaterialConfig) {
```
Comment (arvindsv) [inline config/config-api/src/com/thoughtworks/go/config/materials/MaterialConfigs.java]: Can we put all of these if-else ladders somewhere else? I'm getting worried that there are too many places this is being done. Putting them in one place is not very different, but it's at least very visible where it is being used.

File: config/config-api/src/com/thoughtworks/go/config/materials/ScmMaterialConfig.java [hunk 15]
```diff
@@ -238,29 +238,25 @@ public void setDestinationFolderError(String message) {
...
         }
         File myDir = new File(myDirPath);
-        File otherDir = new File(otherDirPath);
+        File otherDir = new File(otherSCMMaterialFolder);
         try {
             if (FileUtil.isSubdirectoryOf(myDir, otherDir)) {
-                String message = "Invalid Destination Directory.Every material needs a different destination directory and the directories should not be nested.";
-                this.setDestinationFolderError(message);
-                otherMaterial.setDestinationFolderError(message);
+                addError(FOLDER, "Invalid Destination Directory.Every material needs a different destination directory and the directories should not be nested.");
```
Comment (arvindsv) [inline config/config-api/src/com/thoughtworks/go/config/materials/ScmMaterialConfig.java]: If you're making another commit, can you add a space between the two sentences?

File: agent/src/com/thoughtworks/go/agent/AgentController.java (hunk 1 above)
Comment (srinivasupadhya) [inline agent/src/com/thoughtworks/go/agent/AgentController.java]: umm..yes, but I thought having separate args is easier to test interms of mocks etc. Im ok to make it extensions containing all extensions in it.

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java (hunk 6 above)
Comment (srinivasupadhya) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: SCMMaterials as things like `getURL`, `getUserName` etc. which cannot be answered by PluggableSCMMaterial.

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java (hunk 3 above)
Comment (srinivasupadhya) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: I dint want to use PluggedSCMConfig since Im already using PluggableSCMConfig in Pipeline -> Materials. I have kept it SCMs & SCM on same lines as PackageRepositories & PackageRepository. I dont mind changing it to SCMsConfig & SCMConfig.

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java (hunk 4 above)
Comment (srinivasupadhya) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: I thought they would be needed for tests.

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java (hunk 5 above)
Comment (srinivasupadhya) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: on same lines as [packagematerial](https://github.com/gocd/gocd/blob/7958ac8b613031660ee00f0591af792c06c0f74c/common/src/com/thoughtworks/go/config/materials/PackageMaterial.java#L91). I think packagematerial isn't valid until packgedefinition is set & hence the same assumption can be made? @pshivana what do you think?

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java (hunk 7 above)
Comment (srinivasupadhya) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: because super.equals does
```java
        if (name != null ? !name.equals(that.name) : that.name != null) {
            return false;
        }
        if (type != null ? !type.equals(that.type) : that.type != null) {
            return false;
        }
```
type is checked by `getClass() != o.getClass()` while name will be same since its referring to same copy.

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java (hunk 8 above)
Comment (srinivasupadhya) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: fingerprint & scm-id will be same & can be assumed? not sure why different things were used in equals & hashcode. Im ok making them same. @pshivana?

File: common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java (hunk 10 above)
Comment (srinivasupadhya) [inline common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java]: Yes. I need to check if this requires throwing an exception. Also the plugin needs to be initialised with JobConsole so the plugin could log out/err into job console.

File: common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java (hunk 9 above)
Comment (srinivasupadhya) [inline common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java]: not all materials have destination directory. specially DependencyMaterial.

File: config/config-api/src/com/thoughtworks/go/config/materials/MaterialConfigs.java (hunk 14 above)
Comment (srinivasupadhya) [inline config/config-api/src/com/thoughtworks/go/config/materials/MaterialConfigs.java]: Since this is the last change for material Im ok to leave it like this.

File: config/config-api/src/com/thoughtworks/go/config/materials/MaterialConfigs.java (hunk 13 above)
Comment (srinivasupadhya) [inline config/config-api/src/com/thoughtworks/go/config/materials/MaterialConfigs.java]: Ok. I assumed getFolder will be present only in SCMMaterial since the folder support was only in SCM materials. I will change this.

File: config/config-api/src/com/thoughtworks/go/config/materials/ScmMaterialConfig.java (hunk 15 above)
Comment (srinivasupadhya) [inline config/config-api/src/com/thoughtworks/go/config/materials/ScmMaterialConfig.java]: okies.

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java (hunk 5 above)
Comment (pshivana) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: getScmConfig() will not return null here. It was done so in getFingerprint() as part of fixing validations and fixing specs

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java (hunk 8 above)
Comment (pshivana) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: I agree hashCode() and equals() should be done same way, it should be changed for PackageMaterial as well

File: common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java (hunk 11 above)
Comment (pshivana) [inline common/src/com/thoughtworks/go/domain/materials/scm/PluggableSCMMaterialAgent.java]: Yes....plugin api classes should not be leaked beyond plugin access

File: config/config-api/src/com/thoughtworks/go/domain/scm/SCMs.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+    public void removeSCM(String id) {
+        SCM scmToBeDeleted = this.find(id);
+        if (scmToBeDeleted == null) {
+            throw new RuntimeException(String.format("Could not find SCM with id '%s'", id));
+        }
+        this.remove(scmToBeDeleted);
+    }
+
+    private void validateNameUniqueness() {
```
Comment (arvindsv) [inline config/config-api/src/com/thoughtworks/go/domain/scm/SCMs.java]: Do you really need this, since the cruise-config.xsd (below) has a name uniqueness check?

**Comment by srinivasupadhya** [comment]
have added ability to get view template, stored in metadata-store (will be used when UI is built). fixed equals & hashcode, workingDirectory(), folder() & the space in error message.

File: common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java (hunk 8 above)
Comment (srinivasupadhya) [inline common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java]: fixed.

**Comment by arvindsv** [comment]
Sigh. You're not helping this get reviewed, by adding more commits. Especially ones like [this](https://github.com/srinivasupadhya/gocd/commit/e2cfe6a3ad20b9a4da93178e2d9bdbde8e8af696).

File: plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/scm/SCMConfiguration.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+import com.thoughtworks.go.plugin.api.config.Options;
+import com.thoughtworks.go.plugin.api.config.Property;
+
+public class SCMConfiguration implements Comparable {
+
+    public static final Option<Boolean> REQUIRED = Property.REQUIRED;
+    public static final Option<Boolean> PART_OF_IDENTITY = Property.PART_OF_IDENTITY;
+    public static final Option<Boolean> SECURE = Property.SECURE;
+    public static final Option<String> DISPLAY_NAME = Property.DISPLAY_NAME;
+    public static final Option<Integer> DISPLAY_ORDER = Property.DISPLAY_ORDER;
```
Comment (arvindsv) [inline plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/scm/SCMConfiguration.java]: Is the DISPLAY_ORDER needed when the view is coming from a template?

File: server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    @Override
+    public void insertLatestOrNewModifications(final Material material, MaterialInstance materialInstance, File folder, Modifications list) {
+        final PluggableSCMMaterialInstance currentMaterialInstance = (PluggableSCMMaterialInstance) materialInstance;
+
+        final PluggableSCMMaterialInstance latestMaterialInstance = (PluggableSCMMaterialInstance) material.createMaterialInstance();
+        if (currentMaterialInstance.shouldUpgradeTo(latestMaterialInstance)) {
+            transactionTemplate.execute(new TransactionCallback() {
+                @Override
+                public Object doInTransaction(TransactionStatus transactionStatus) {
+                    PluggableSCMMaterialInstance materialInstance = (PluggableSCMMaterialInstance) materialRepository.find(currentMaterialInstance.getId());
```
Comment (arvindsv) [inline server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java]: This should be moved into MaterialRepository. If any kind of caching considerations are applicable here (now or sometime later), then, no one would look at this usage, here. I mean, if the material instance is cached somewhere (usually in MaterialRepository), it'll need to be removed from cache. That belongs in MaterialRepository, not here.

File: server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+                    materialInstance.upgradeTo(latestMaterialInstance);
+                    materialRepository.saveOrUpdate(materialInstance);
+                    return materialInstance;
+                }
+            });
+        }
+        scmMaterialUpdater.insertLatestOrNewModifications(material, currentMaterialInstance, folder, list);
```
Comment (arvindsv) [inline server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java]: Once upgraded, is currentMaterialInstance the right one to be used? Is that because the actual object of currentMaterialInstance will be updated in line 55?

**Comment by arvindsv** [comment]
Alright, I think I'm done. I have nothing else at this point. Only three more questions for you to answer!

File: plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/scm/SCMConfiguration.java (hunk 17 above)
Comment (srinivasupadhya) [inline plugin-infra/go-plugin-access/src/com/thoughtworks/go/plugin/access/scm/SCMConfiguration.java]: yes. `DISPLAY_NAME` & `DISPLAY_ORDER` will be used in material listing page to list properties & their values.

File: server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java (hunk 18 above)
Comment (srinivasupadhya) [inline server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java]: on same lines as [this](https://github.com/gocd/gocd/blob/7958ac8b613031660ee00f0591af792c06c0f74c/server/src/com/thoughtworks/go/server/materials/PackageMaterialUpdater.java#L50). @pshivana what do you think?

File: server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java (hunk 19 above)
Comment (srinivasupadhya) [inline server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java]: on same lines as [this](https://github.com/gocd/gocd/blob/7958ac8b613031660ee00f0591af792c06c0f74c/server/src/com/thoughtworks/go/server/materials/PackageMaterialUpdater.java#L50). @pshivana what do you think?

File: server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java (hunk 18 above)
Comment (pshivana) [inline server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java]: Yes it can be moved to MaterialRepository

File: server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java (hunk 19 above)
Comment (pshivana) [inline server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java]: I don't it matters which instance being used on line 61. It should work either way.

Comment (srinivasupadhya) [inline server/src/com/thoughtworks/go/server/materials/PluggableSCMMaterialUpdater.java]: ok will move both usages (packge-repository & pluggable-scm) to material repository in next PR.
also, will make equals & hashcode for packagematerial consistent in that PR.

**Comment by srinivasupadhya** [comment]
ok, i am done too. thanks @arvindsv & @pshivana for the review.
@arvindsv sorry for adding new commit in middle of the review. it was required for UI & dint want to wait for new PR.

**Comment by arvindsv** [comment]
> ok will in next PR:

> move both usages (packge-repository & pluggable-scm) to material repository
> will make equals & hashcode for packagematerial consistent
> make PackgeAsMaterialExtension, SCMExtension & TaskExtension into Extensions on the agent side.
> see if name uniqueness validity is required, both in packge-repository & pluggable-scm

This should be done too.
